internal/role/p2p/k0s.go

```diff
@@ -0,0 +1,210 @@
+	providerConfig "github.com/kairos-io/provider-kairos/v2/internal/provider/config"
+	service "github.com/mudler/edgevpn/api/client/service"
+)
+
+const (
+	K0sMasterName = "controller"
+	K0sWorkerName = "worker"
+)
+
+type K0sNode struct {
+	providerConfig *providerConfig.Config
+	roleConfig     *service.RoleConfig
+	ip             string
+	role           string
+}
+
+func (k *K0sNode) IsWorker() bool {
+	return k.role == RoleWorker
+}
+
+func (k *K0sNode) K8sBin() string {
+	return utils.K0sBin()
+}
+
+func (k *K0sNode) DeployKubeVIP() error {
+	pconfig := k.ProviderConfig()
+	if pconfig.KubeVIP.IsEnabled() {
+		return errors.New("KubeVIP is not yet supported with k0s")
+	}
+
+	return nil
+}
+
+func (k *K0sNode) GenArgs() ([]string, error) {
+	var args []string
+
+	pconfig := k.ProviderConfig()
+	// if pconfig.P2P.UseVPNWithKubernetes() {
+	// k3s uses this check to connet using the interface; however k0s does not have a way to configure that
+	// }
+
+	if pconfig.KubeVIP.IsEnabled() {
+		return args, errors.New("KubeVIP is not yet supported with k0s")
+	}
+
+	if pconfig.P2P.Auto.HA.ExternalDB != "" {
+		return args, errors.New("ExternalDB is not yet supported with k0s")
+	}
+
+	if k.HA() && !k.ClusterInit() {
+		return args, errors.New("HA is not yet supported with k0s")
+	}
+
+	args = pconfig.K0s.AppendArgs(args)
+
+	return args, nil
+}
+
+func (k *K0sNode) EnvUnit() string {
+	return machine.K0sEnvUnit("k0scontroller")
+}
+
+func (k *K0sNode) Service() (machine.Service, error) {
+	if k.IsWorker() {
+		return machine.K0sWorker()
+	}
+
+	return machine.K0s()
+}
+
+func (k *K0sNode) Token() (string, error) {
+	if k.IsWorker() {
+		return k.RoleConfig().Client.Get("workertoken", "token")
+	}
+
+	return k.RoleConfig().Client.Get("controllertoken", "token")
+}
+
+func (k *K0sNode) GenerateEnv() (env map[string]string) {
+	env = make(map[string]string)
+
+	if k.HA() && !k.ClusterInit() {
+		nodeToken, _ := k.Token()
+		env["K0S_TOKEN"] = nodeToken
+	}
+
+	pConfig := k.ProviderConfig()
+
+	if pConfig.K0s.ReplaceEnv {
+		env = pConfig.K0s.Env
+	} else {
+		// Override opts with user-supplied
+		for k, v := range pConfig.K0s.Env {
+			env[k] = v
+		}
+	}
+
+	return env
+}
+
+func (k *K0sNode) SetProviderConfig(c *providerConfig.Config) {
+	k.providerConfig = c
+}
+
+func (k *K0sNode) ProviderConfig() *providerConfig.Config {
+	return k.providerConfig
+}
+
+func (k *K0sNode) SetRoleConfig(c *service.RoleConfig) {
+	k.roleConfig = c
+}
+
+func (k *K0sNode) RoleConfig() *service.RoleConfig {
+	return k.roleConfig
+}
+
+func (k *K0sNode) HA() bool {
+	return k.role == RoleMasterHA
+}
+
+func (k *K0sNode) ClusterInit() bool {
+	// k0s does not have a cluster init role like k3s. Instead we should have a way to set in the config
+	// if the user wants a single node cluster, multi-node cluster, or HA cluster
+	return false
+}
+
+func (k *K0sNode) IP() string {
+	return k.ip
+}
+
+func (k *K0sNode) propagateData() error {
```

Comment:
This method is selectively permissive but I don't understand why. It tries to create both a controller and worker token (the node is both a master and a worker?). Also it only prints the errors without failing. Why do we expect it it work if the token generation failed? Isn't it better if we return in that case?

Reply:
yup indeed, it basically is waiting in the background for the k0scontroller service to run, before that, both calls to create the token will fail but it doesn't necessarily mean that there's an issue but it might just be that it's still waking up

and yes we do both tokens because we don't know who will try to connect on the other side, either a worker or another controller node so we need to propagate both on the network. At least that's the logic I understand from k0s docs, but last night that I was checking the tokens they seemed pretty much the same, but maybe it's just because of my conf or because it was pretty late 🤣